# Data Audit - sales_data_raw.xlsx
## Project: Product Placement Optimisation
## Student: Samikshya Baniya
## Date: June 2026
### Purpose: Examine raw data without modifying it. Document all findings.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
print("Libraries loaded successfully!")
print("Starting Data Audit for sales_data_raw.xlsx")


Libraries loaded successfully!
Starting Data Audit for sales_data_raw.xlsx


## Section 1: Loading the Raw Data

In [3]:
df = pd.read_excel(
    r'D:\softwarica\Sem 6\Individual Project\product-placement-optimization\data\raw\sales_data_raw.xlsx',
    header=6
)

print("Data loaded successfully!")
print(f"Shape: {df.shape}")

Data loaded successfully!
Shape: (768223, 14)


### First look at the data

In [4]:
df.head()

,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9,PW Term,Unnamed: 11,BW Term,Unnamed: 13
0,Date,Miti,Inv.No,Customer,Product Group,Product,Uom,Qty,Rate,B. Amount,DISCOUNT,VAT,ROUND OFF,Total Amount
1,2025-07-17 00:00:00,01/04/2082,ABT-000020,CASH A/C,POOJA ITEMS,Til Oil 500ml,PCS,1,175,154.867256,0,20.132744,0,175
2,2025-07-17 00:00:00,01/04/2082,ABT-000029,CASH A/C,GAULE,Gaule Chukamilo,PCS,1,105,105,0,0,0,105
3,2025-07-17 00:00:00,01/04/2082,ABT-000029,CASH A/C,ESSENTIAL FOOD ITEM,Krishma Mixnut 100gm,PCS,1,80,70.79646,0,9.20354,0,80
4,2025-07-17 00:00:00,01/04/2082,ABT-000029,CASH A/C,ESSENTIAL FOOD ITEM,Mong Khosta,KG,4,185,740,0,0,0,740


### Column Names and Data Types

In [5]:
print("Column Names:")
print("---")
print(df.columns.tolist())
print("---")
print("\nData Types:")
print("---")
print(df.dtypes)
print("---")
print(f"\nTotal Rows: {df.shape[0]:,}")
print(f"Total Columns: {df.shape[1]}")

Column Names:
---
['Unnamed: 0', 'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4', 'Unnamed: 5', 'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9', 'PW Term', 'Unnamed: 11', 'BW Term', 'Unnamed: 13']
---

Data Types:
---
Unnamed: 0     object
Unnamed: 1        str
Unnamed: 2        str
Unnamed: 3        str
Unnamed: 4        str
Unnamed: 5        str
Unnamed: 6        str
Unnamed: 7     object
Unnamed: 8     object
Unnamed: 9     object
PW Term        object
Unnamed: 11    object
BW Term        object
Unnamed: 13    object
dtype: object
---

Total Rows: 768,223
Total Columns: 14


In [6]:
df_peek = pd.read_excel(
    r'D:\softwarica\Sem 6\Individual Project\product-placement-optimization\data\raw\sales_data_raw.xlsx',
    header=None,
    nrows=15
)
print(df_peek)

                                                   0           1           2   \
0   Exported   Miti/Time : 06/02/2083( 20/05/2026 ...         NaN         NaN   
1                    BANIYA SHOPPING CENTER  PVT.LTD.         NaN         NaN   
2                                LAMACHOUR-16 POKHARA         NaN         NaN   
3             PAN No : 610418608 ; Phone : 061-443259         NaN         NaN   
4                              Sales Register Detail          NaN         NaN   
5   From 01/04/2082 ( 17/07/2025)  To  06/02/2083 ...         NaN         NaN   
6                                                 NaN         NaN         NaN   
7                                                Date        Miti      Inv.No   
8                                 2025-07-17 00:00:00  01/04/2082  ABT-000020   
9                                 2025-07-17 00:00:00  01/04/2082  ABT-000029   
10                                2025-07-17 00:00:00  01/04/2082  ABT-000029   
11                          

In [7]:
df = pd.read_excel(
    r'D:\softwarica\Sem 6\Individual Project\product-placement-optimization\data\raw\sales_data_raw.xlsx',
    header=7
)

print("Data loaded successfully!")
print(f"Shape: {df.shape}")

Data loaded successfully!
Shape: (768222, 14)


## Section 2: Column Names and Data Types

In [8]:
print("COLUMN NAMES:")
print("---")
print(df.columns.tolist())
print("---")
print("\nDATA TYPES:")
print("---")
print(df.dtypes)

COLUMN NAMES:
---
['Date', 'Miti', 'Inv.No', 'Customer', 'Product Group', 'Product', 'Uom', 'Qty', 'Rate', 'B. Amount', 'DISCOUNT', 'VAT', 'ROUND OFF', 'Total Amount']
---

DATA TYPES:
---
Date             datetime64[us]
Miti                        str
Inv.No                      str
Customer                    str
Product Group               str
Product                     str
Uom                         str
Qty                     float64
Rate                    float64
B. Amount               float64
DISCOUNT                float64
VAT                     float64
ROUND OFF               float64
Total Amount            float64
dtype: object


### Findings: Column Names and Data Types

- 14 columns total
- Date is datetime64 - ready for time analysis
- Miti is kept as string - Nepali calendar not supported by Python datetime
- Qty, Rate, B.Amount, VAT, ROUND OFF, Total Amount are float64 - ready for maths
- DISCOUNT is float64 but we will check if it has any real values
- No data type conversions needed at this stage

## Section 3: Key Numbers

In [9]:
print("KEY NUMBERS")
print("-" * 50)
print(f"Total rows:                 {len(df):,}")
print(f"Total unique invoices:      {df['Inv.No'].nunique():,}")
print(f"Total unique products:      {df['Product'].nunique():,}")
print(f"Unique product groups:      {df['Product Group'].nunique():,}")
print(f"Unique customers:           {df['Customer'].nunique():,}")
print(f"Date range:                 {df['Date'].min().date()} to {df['Date'].max().date()}")

KEY NUMBERS
--------------------------------------------------
Total rows:                 768,222
Total unique invoices:      218,037
Total unique products:      5,681
Unique product groups:      38
Unique customers:           25
Date range:                 2025-07-17 to 2026-05-20


## Section 4: Product Group Audit

In [10]:
print("ALL PRODUCT GROUPS BY COUNT:")
print("-" * 70)
for i, (group, count) in enumerate(df['Product Group'].value_counts().items(), 1):
    pct = (count / len(df)) * 100
    print(f"{i:2d}. {group:40s} | {count:6,} rows | {pct:4.1f}%")

ALL PRODUCT GROUPS BY COUNT:
----------------------------------------------------------------------
 1. ESSENTIAL FOOD ITEM                      | 170,669 rows | 22.2%
 2. CANNED AND PACK GOODS                    | 96,889 rows | 12.6%
 3. SOAP AND CLEANER                         | 50,868 rows |  6.6%
 4. BISCUITS                                 | 47,062 rows |  6.1%
 5. CHOCOLATE & CANDY                        | 44,774 rows |  5.8%
 6. TEA & MASALA ITEM                        | 44,202 rows |  5.8%
 7. Cooking Oil                              | 37,265 rows |  4.9%
 8. HOUSEHOLD ITEM                           | 36,091 rows |  4.7%
 9. NOODLES ITEMS                            | 31,964 rows |  4.2%
10. SOFT DRINK                               | 30,832 rows |  4.0%
11. PERSONAL CARE                            | 29,669 rows |  3.9%
12. DAIRY PRODUCT                            | 29,506 rows |  3.8%
13. Rice                                     | 16,123 rows |  2.1%
14. Cigarette               

In [11]:
print("ALL PRODUCT GROUPS (alphabetical - to spot duplicates):")
print("-" * 60)
for group in sorted(df['Product Group'].dropna().unique()):
    print(f"  '{group}'")

ALL PRODUCT GROUPS (alphabetical - to spot duplicates):
------------------------------------------------------------
  'BABY CARE'
  'BISCUITS'
  'Bakery Item'
  'Battisa Masala'
  'Birthday Item'
  'CANNED AND PACK GOODS '
  'CANNED AND PACK ITEM'
  'CEREALS'
  'CHOCOLATE & CANDY'
  'COFFEE ITEMS'
  'Cigarette'
  'Cooking Oil '
  'DAIRY PRODUCT'
  'DRY FRUIT ITEMS '
  'ESSENTIAL FOOD ITEM'
  'Electrical Supplies '
  'FROZEN'
  'GAULE'
  'HEALTH & WELLNESS'
  'HOUSEHOLD ITEM'
  'Hard Drinks'
  'KOREAN ITEMS'
  'LAXMI DALMOT'
  'Meat and seafood'
  'NOODLES ITEMS '
  'PAICHO ITEMS '
  'PERSONAL CARE'
  'PERSONAL_CARE'
  'PET CARE'
  'POOJA ITEMS'
  'Rice'
  'SOAP AND CLEANER'
  'SOFT DRINK'
  'STATIONARY ITEM'
  'SWEETS '
  'Surti'
  'TEA & MASALA ITEM'
  'Vegetables and Fruits'


In [12]:
pg = df['Product Group'].dropna().astype(str)
space_issues = (pg != pg.str.strip()).sum()
print(f"Product Group values with trailing spaces: {space_issues:,}")

prod = df['Product'].dropna().astype(str)
prod_space = (prod != prod.str.strip()).sum()
print(f"Product values with trailing spaces:       {prod_space:,}")

Product Group values with trailing spaces: 168,864
Product values with trailing spaces:       1,216


### Findings: Product Group Audit

- 38 unique product groups found
- Inconsistent capitalization: mix of UPPER CASE and Title Case
- Trailing spaces found in 168,864 Product Group values (22% of data)
- Trailing spaces found in 1,216 Product values
- Duplicate categories found:
    - PERSONAL CARE and PERSONAL_CARE are the same category
    - CANNED AND PACK GOODS and CANNED AND PACK ITEM are the same category
- All issues will be fixed in 02_data_cleaning.ipynb

## Section 5: Numeric Column Audit

In [13]:
numeric_cols = ['Qty', 'Rate', 'B. Amount', 'VAT', 'Total Amount']

print("NUMERIC COLUMN SUMMARY")
print("-" * 60)
df[numeric_cols].describe().T.round(2)

NUMERIC COLUMN SUMMARY
------------------------------------------------------------


,count,mean,std,min,25%,50%,75%,max
Qty,768221.0,1.89,12.02,0.05,1.00,1.00,1.00,1900.00
Rate,768221.0,243.64,499.77,0.28,65.00,110.00,220.00,30060.18
B. Amount,768221.0,258.88,522.62,0.35,75.22,131.86,243.36,34605.28
VAT,768064.0,25.48,58.34,0.00,3.45,11.50,26.46,4498.69
Total Amount,768221.0,284.27,569.57,0.40,80.00,140.00,270.00,39104.00


In [14]:
print(f"Rows with Qty <= 0:          {(df['Qty'] <= 0).sum():,}")
print(f"Rows with Rate <= 0:         {(df['Rate'] <= 0).sum():,}")
print(f"Rows with Total Amount <= 0: {(df['Total Amount'] <= 0).sum():,}")

Rows with Qty <= 0:          0
Rows with Rate <= 0:         0
Rows with Total Amount <= 0: 0


### Findings: Numeric Column Audit

- All numeric columns have correct float64 data type
- Qty minimum is 0.05 - weighed items sold in fractions
- Qty maximum is 1900 - bulk purchases exist
- Total Amount maximum is Rs 39,104 - likely bulk business purchase
- VAT has 768,064 values but others have 768,221 - 157 rows missing VAT
- No negative values found in Qty, Rate or Total Amount
- No data type fixes needed

## Section 6: Discount, VAT and Round Off Audit

In [15]:
for col in ['DISCOUNT', 'VAT', 'ROUND OFF']:
    print(f"\n{col} AUDIT")
    print("-" * 50)
    print(f"Null values:     {df[col].isna().sum():,}")
    print(f"Zero values:     {(df[col] == 0).sum():,}")
    print(f"Positive values: {(df[col] > 0).sum():,}")
    print(f"Maximum value:   {df[col].max()}")


DISCOUNT AUDIT
--------------------------------------------------
Null values:     158
Zero values:     759,261
Positive values: 0
Maximum value:   0.0

VAT AUDIT
--------------------------------------------------
Null values:     158
Zero values:     158,044
Positive values: 610,020
Maximum value:   4498.6864000000005

ROUND OFF AUDIT
--------------------------------------------------
Null values:     158
Zero values:     744,282
Positive values: 10,658
Maximum value:   0.5
